File: 백준/Gold/2118. 두 개의 탑/두 개의 탑.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int solve(int n, const vector<int>& lst){
    int lo = 0;
    int hi = 1;
    int rtn = 0;
    int clockwise, counterclockwise;
    
    while (hi < n){
        clockwise = lst[hi] - lst[lo];
        counterclockwise = (lst[n] - lst[hi]) + lst[lo];

        rtn = max(rtn, min(clockwise, counterclockwise));
        
        if (counterclockwise > clockwise) hi++;
        else lo++;
    }

    return rtn;
}

int main() {
    ios_base::sync_with_stdio(false); // 입출력 속도 향상
    cin.tie(NULL);

    int n;
    cin >> n;

    vector<int> lst(n+1);
    lst[0] = 0;

    for (int i = 1; i <= n; i++){
        cin >> lst[i];
        lst[i] += lst[i-1];
    }

    int result = solve(n, lst);
    cout << result << endl;
    
    return 0;
}

File: 백준/Gold/2118. 두 개의 탑/두 개의 탑.js
const readline = require('readline');

const rl = readline.createInterface({
  input : process.stdin,
  output : process.output
})

let input = [];

rl.on('line', (line)=>{
  input.push(Number(line));
  if (input[0] < input.length) rl.close();
})

rl.on('close', ()=>{
  const N = input[0];
  const arr = input.slice(1);
  const result = solve(N, arr);
  console.log(result);
})


const solve = (n, arr) => {
  let rtn = 0;
  for (let i = 1; i<n; i++) arr[i] += arr[i-1]; // 누적합
  arr.unshift(0)
  let lo = 0;
  let hi = 1;
  let clockwise, counterclockwise;
  while (lo < hi){
    clockwise = arr[hi] - arr[lo];
    counterclockwise = (arr[n] - arr[hi]) + arr[lo];
    rtn = Math.max(rtn, Math.min(clockwise, counterclockwise));

    if (counterclockwise > clockwise) hi++;
    else lo++;

    if (hi >= n) break;
  }

  return rtn;
}
